pdd/PDD-multistream.md

```diff
@@ -0,0 +1,139 @@
+# PDD for multistream (example/use case)
+
+## Protocol
+
+The motivation and base of the protocol can be found here:
+
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream---self-describing-protocol-stream
+- https://github.com/ipfs/specs/blob/wire/protocol/network/wire.md#multistream-selector---self-describing-protocol-stream-selector
+
+The multistream protocol does not cover:
+
+- discovering participants, selecting transports, and establishing connections
+- managing the state of the connection
+
+multistream enables several types of streams to be used over one single stream, like an intelligent message broker that offers the ability to negotiate the protocol and version that is going to be used. To simplify, a visual representation can be:
+
+```
+┌ ─ ─ ─ ─ ─ ─ ┌ ─ ─ ─ ─ ─ ─ ┐┌ ─ ┐
+ dht-id/1.0.1│ bitswap/1.2.3  ...
+└ ─ ─ ─ ─ ─ ─ └ ─ ─ ─ ─ ─ ─ ┘└ ─ ┘
+┌────────────────────────────────┐
+│ multistream-select             │
+└────────────────────────────────┘
+┌────────────────────────────────┐
+│transport                       │
+│  [TCP, UDP, uTP, ...]          │
+└────────────────────────────────┘
+```
+
+multistream doesn't cover stream multiplexing over the same connection, however, we can achieve this by leveraging functionality offered by SPDY or HTTP/2.
+
+```
+┌─── ───┐┌─── ───┐┌ ┌─── ───┐┌─── ───┐┌┐
+ a/1.0.0  b/1.0.0  │ a/1.0.0  b/1.0.0 ││
+└─── ───┘└─── ───┘└ └─── ───┘└─── ───┘ ┘
+┌─── ──── ──── ──── ┌─── ──── ──── ──── ┌───
+│multistream-select││multistream-select││...│
+└ ──── ──── ──── ──┘└ ──── ──── ──── ──┘└ ──┘
+┌─── ──── ──── ──── ──── ──── ──── ──── ──── ┌───
+│stream multiplexing                        ││...│
+│ [HTTP/2, SPDY]                            ││   │
+└─ ──── ──── ──── ──── ──── ──── ──── ──── ─┘└─ ─┘
+┌────────────────────────────────────────────────┐
+│ multistream-select                             │
+└────────────────────────────────────────────────┘
+┌────────────────────────────────────────────────┐
+│transport                                       │
+│  [TCP, UDP, uTP, ...]                          │
+└────────────────────────────────────────────────┘
+```
+
+multistream messages have the following structure:
+
+```
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+ ┌───────────────┐┌───────┐│
+││varint with    ││message│
+ │ message length││       ││
+│└───────────────┘└───────┘
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
```

Comment:
> one comment: we don't need this frame for every message.
> ... (maybe you already mean this -- wasnt getting it from the above)

Yes, I hope now it's more clear, once a protocol is handshake'd, it is up to that protocol to control the stream and encode however they want.